src/Confluent.Kafka/Producer.cs

```diff
@@ -665,53 +666,96 @@ internal Producer(ProducerBuilder<TKey, TValue> builder)
             Message<TKey, TValue> message,
             CancellationToken cancellationToken = default(CancellationToken))
         {
-            var keyBytes = (keySerializer != null)
```

Comment:
there are a bunch of changes to this method which mean that any error encountered due to a call to produce will result in a `ProduceException<K,V>` (note: which has a `DeliveryResult<K,V>` property). Formerly, a `ProduceException` would only occur on receiving a delivery report with an error (in the `Task` based `ProduceAsync` case only). Now, this will also be generated when a produce call returns an error immediately, or due to a serialization problem (refer to discussion in consume case).

This is more convenient. Imagine we have a code block with a bunch of `async` produce calls and some other logic that could also throw. We could wrap it in a `try`/`catch` block, and catch a `ProduceException<K,V>`. We know that if a produce call failed for any reason, we'll find out about it here. We'll also have the context (i.e. topic, key, value etc) regardless of whether the failure was immediate or not. Before this change, we'd need to catch the `KafkaException` type as well + whatever we expect the serializers could throw (+ we'd need to know what that is!) to get every case + in the immediate cases the exception wouldn't include context, which is useful.

Note: we could add a non-generic typed variant of `ProduceException` to allow exceptions from different typed producers to be caught at once. can be added without breaking the API, so left as a todo.

Reply:
exception to the above: programmer error type exceptions (eg argument exception) are better if not coerced into a ProduceException.

Reply:
> Note: we could add a non-generic typed variant of `ProduceException` to allow exceptions from different typed producers to be caught at once. can be added without breaking the API, so left as a todo.

That answers my previous concern :+1:

Reply:
well you can already do this by catching `KafkaException` - only need `ProduceException<K,V>` if you need access to the info it provides. `ProduceException` is kind of halfway in between (doesn't provide K,V).